docs/proposals/push-metrics.md

```diff
@@ -0,0 +1,69 @@
+interface for pulling from other sources.  However, in certain cases, it is
+more useful to be able to have services push metrics into Heapster, instead
+of having Heapster pull metrics.
+
+For instance, supporting push metrics makes it easy for cluster admins to add
+custom metrics into Heapster with relatively minimal effort: they can simply
+write a program or script which collects and processes the information, and
+then add a recurring Job or cron job which pushes the metrics (instead of
+having to write a source for Heapster).  This also enables existing tooling
+designed with the push model in mind to be adapted to provide metrics through
+Heapster.
+
+## Design
+
+### API
+
+Metrics added through push metrics will be available as normal through the
+Heapster model.
+
+To add metrics, metrics producers will `POST` new metrics to
+`/api/v1/push/{format}` in the format specified by `{format}`.  Producers will
+be authenticated similarly to the way Heapster currently authenticates clients:
+producers will present a certificate signed by a CA, and Heapster will be
+configurable to only allow certain names to push metrics.
+
+Pushed metrics will then be additionally scoped by producer name, to avoid
+conflicts.
+
+### Metrics Format
+
+The underlying design can support multiple format "backends".  Initially, only
+the prometheus text format will be supported at `/api/v1/push/prometheus`.  The
+metric name in the prometheus metrics will be mapped to a custom metric with
+the name `{client_name}/{metric_name}` in Heapster (with the client name coming
+being the same as used in authentication), and the following labels will be
+used to determine which object a metric is associated with:
+
+- For kubernetes-related metrics, the `namespace` label will indicate
+  namespace, the `pod` label will indicate pod, and the `container` label will
+  indicate container.  If only `namespace` is present, then the metric will be
+  considered namespace-level.  If `namespace` and `pod` are present, then the
+  metric will be considered pod-level.  If all three are present, the metric
+  will be considered container-level.
+
+- For non-kubernetes-related metrics, the `node` label will indicate node, and
+  the `container` label will indicate free container.   If only `node` is
+  present, the metric will be considered node-level.  Otherwise, the metric
+  will be considered free-container-level.
+
+Any additional labels will be treated the same as labels on existing custom
+metrics (currently multiple custom metrics with the same name, but different
+labels, are ignored, but this seems like an oversight and should probably be
+fixed).
+
+Metric lines must have timestamps, which will be used as the timestamp
+associated with the metric in Heapster (similarly to how pull metrics use the
+timestamp provided by Kubelet).  As normal, this requires the producer's clock
+to be in sync to yield reasonable results.
+
+## Open Questions
```

Comment:
Metrics pushers can't overwrite/interfere with each other, since each metric is stored as prefixed by pusher's name, as determined during the auth step, unless you had multiple sources that shared the same name but were not otherwise coordinated.  That scenario seems like something that should just be avoided in this case.  I can make that explicit if need be. 

In general, I imagined that the use cases for this are more designed for either cluster-admin-created components, or admin-blessed components, and thus some of this would be a case of "does the admin trust themselves".

Nonetheless, I could be desirable to limit metrics per pusher, pushes per second, etcd.  It would be fairly easy to do so on a global scale, but I could see a case for having per-pusher controls on those variables.  The question becomes how complicated we want to make this.